common/sampling.h

```diff
@@ -36,7 +36,7 @@ struct common_sampler;
 
 // llama_sampler API overloads
 
-struct common_sampler * common_sampler_init(const struct llama_model * model, const struct common_sampler_params & params);
+struct common_sampler * common_sampler_init(const struct llama_model * model, const struct common_sampler_params & params, int32_t context_size);
```

Comment:
Instead of adding this context size in so many places, better to use `llama_n_ctx_train(model)` when initializing the sampler. This value will always be larger or equal to the current context size, so it would be safe to use as a default value when DRY n_repeat is not specified. This way we avoid complicating the public interfaces with something very specific to the DRY sampler.

Reply:
@ggerganov I tried this running llama-server with `Mistral-Nemo-Instruct-2407-Q5_K_S.gguf` and CLI arguments ` -c 32000 -ngl 41 -b 2048 -t 6 -tb 12` and got a value that was higher by a factor of 32:
```
slot         init: id  0 | task -1 | new slot n_ctx_slot = 32000
... skipping some lines ...
Comparison between context_size values, old passed context size is: 32000
New calculated llama_n_ctx_train context size is: 1024000
```
Are we ok with it being that much higher?

Reply:
In looking further into this, I now see that this comes from hparams, so the code would allocate last_tokens and dry_repeat_count based on this theoretical max value for the model. Is that ok?

Reply:
The last time I coded in C++ prior to these PRs was in the 90s (wasn't kidding when I said I was rusty!), but I am probably too stuck in the old-school fashion of counting every kb of RAM used, so I just went ahead and implemented the llama_n_ctx_train(model) change as suggested. :sweat_smile: This restores the simplified common_sampler_init function signature.